streams/api/src/main/java/org/eclipse/microprofile/reactive/streams/spi/Stage.java

```diff
@@ -360,6 +360,34 @@ public FlatMap(Function<?, Graph> mapper) {
     }
   }
 
+  /**
+   * A flat map stage similar to {@link FlatMap} but the mapper function returns a {@link Publisher} instead of a
+   * {@link Graph}.
+   * <p>
+   * The flat map stage should execute the given mapper on each element, and concatenate the publishers emitted by
+   * the mapper function into the resulting stream.
+   * <p>
+   * The graph emitted by the mapper function is guaranteed to have an outlet but no inlet.
+   * <p>
+   * The engine must be careful to ensure only one publisher emitted by the mapper function is running at a time.
+   */
+  final class FlatMapPublisher implements Inlet, Outlet {
+      private final Function<?, Publisher<?>> mapper;
+
+      public FlatMapPublisher(Function<?, Publisher<?>> mapper) {
+          this.mapper = mapper;
+      }
+
+      /**
+       * The mapper function.
+       *
+       * @return The mapper function.
+       */
+       public Function<?, Publisher<?>> getMapper() {
+            return mapper;
+        }
+  }
+
```

Comment:
You actually haven't used this in the `flatMapPublisher` method, you've just wrapped it in a `Stage.FlatMap` which wraps the returned `Publisher` in a `PublisherBuilder`. I think that's the right way to go anyway, I don't think there's any value in having a separate stage for a straight publisher, so this can be deleted.

Reply:
You are absolutely right, and the implementation does not use the stage anyway. I just removed it.

BTW, about the implementation, do you prefer chained `andThen` or a single function?

Reply:
If it were 5 `andThen` methods, I'd say consolidate (performance may start to matter since `andThen` with method references doesn't allow any inlining), but just two it's much of a muchness in my opinion.